#1612 comment children links

BE-Webdesign opened the pull request:
Fixes #1612,

This patch adds a `children` link to the response when the comment has children.  The link is to a collection of comments who share the same parent as the original comment resource.  This enables a client to then recursively move through the links, with pagination parameters to boot, to performantly collect all threaded comments of a single comment.

BE-Webdesign committed:
Fixes #1612 by adding link to collection sharing comment parent.

Fixes #1612,

This patch adds a `children` link to the response when the comment has
children.  The link is to a collection of comments who share the same
parent as the original comment resource.  This enables a client to then
recursively move through the links, with pagination parameters to boot,
to performantly collect all threaded comments of a single comment.

BE-Webdesign commented:
The test cases are not ideal, but it would involve refactoring existing test cases.  I decided to just keep this PR clean and to the point of the issue.  I will open up a separate PR at some point to improve the test suite.  I believe an issue is already open for that and I will try to hunt it down.

kadamwhite commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -640,6 +640,18 @@ protected function prepare_links( $comment ) {
 		}
 
+		// Only grab one comment to verify the comment has children.
+		$comment_children = $comment->get_children( array( 'number' => 1 ) );
```
In code review @rmccue identified this to be a performance concern, as `get_children` runs a query every time that is not object-cached. Performance impact needs to be assessed on a site with a high volume of comments before this can be merged.

BE-Webdesign replied:
I tested this on 1,100,000 comments, using Redis it dropped the response time back to normal.  The query is object cached, so to my understanding it should not be a big problem.  There is very little difference in response time for that magnitude of comments when the children are grabbed or not grabbed. On average over 100 requests each the difference was 2ms.

rmccue replied:
Since we only care about the count, we should pass `count => true`, which queries for `COUNT(*)` instead of the fields. This is less expensive when we actually need to hit the database, and should be equivalent performance for the cached case.

BE-Webdesign replied:
Sounds good, I will check that out.

BE-Webdesign replied:
That looks like it will save even more execution time and the cache will be a lot smaller 👍! Thank you for the tip.

BE-Webdesign commented:
As noted in a Slack conversation today. Performance benchmarks will need to occur to ensure this is fast enough on higher scale installs.  Testing at 10k comments and 100k comments and looking at the differences.  Will be posted at some point soonish.

BE-Webdesign commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -640,6 +640,18 @@ protected function prepare_links( $comment ) {
+		$comment_children = $comment->get_children( array( 'number' => 1 ) );
+		if ( ! empty( $comment_children ) ) {
+			$args = array( 'parent' => $comment->comment_ID );
+			$rest_url = add_query_arg( $args, rest_url( $this->namespace . $this->rest_base ) );
```
This is not the right link it needs a `/` before `$this->rest_base`.

BE-Webdesign commented:
I tested this on 1,100,000 comments, using Redis it dropped the response time back to normal. The query is object cached, so to my understanding it should not be a big problem. There is very little difference in response time for that magnitude of comments, when the children are grabbed or not grabbed. On average over 100 requests for each one with comment children one without, the difference was 2ms.

BE-Webdesign committed:
Fix Link

The link was actually broken oops!!!! Good thing it wasn't merged.  From
a performance perspective the call to `WP_Comment::get_children()` is
object cached from what I can tell.

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into #1612-comment-children-links

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into #1612-comment-children-links

BE-Webdesign committed:
Add in count parameter.

Adding in the count query parameter.  As rmccue pointed out, it is
better to query for the count as opposed to chaching the actual fields
into the object cache, because all we really need is the count.

BE-Webdesign commented:
I added in the count parameter so the object cache will be a lot smaller.

joehoyle commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -646,6 +646,18 @@ protected function prepare_links( $comment ) {
+
+			$links['children'] = array(
+				'href'       => $rest_url,
+				'embeddable' => true,
```
Hmm I'm not sure `embeddable =>true` is a good idea, typically we don't have potentially large lists as `embeddable => true` for example, posts / users lists.

BE-Webdesign replied:
Sounds good, I will change it. The original ticket I believe wanted it to be embeddable but it doesn't make much sense to embed a large collection, especially if we make a batch endpoint soon.

BE-Webdesign committed:
Remove Embeddability

Generally Embeddable links should not be collections or big amounts of
data.

rachelbaker commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -646,6 +646,17 @@ protected function prepare_links( $comment ) {
+			$rest_url = add_query_arg( $args, rest_url( $this->namespace . '/' . $this->rest_base ) );
+
+			$links['children'] = array(
+				'href'       => $rest_url,
```
@BE-Webdesign indentation is off after https://github.com/WP-API/WP-API/pull/2662/commits/b10fb736dff1ecb7addad4c91f24fccb6986c8a6

rachelbaker reviewed (changes requested):
Minor indentation fix needed.

joehoyle commented:
@BE-Webdesign any update here? would be nice to get this out today!

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into #1612-comment-children-links

BE-Webdesign committed:
Fixing Indentation

Thank you @rachelbaker that was a good catch.  The indentation was off
from when embeddable parameter was removed.  This patch fixes the
spacing.

BE-Webdesign commented:
I fixed the indentation, removed its embedability, and changed the query to only run the count for performance reasons.  It should be good to go. The PR still needs the thumbs up from @rmccue and @rachelbaker.  If it is decided that this feature is not needed, then feel free to close the PR.
